### PR TITLE
Loc for magic version and number check

### PR DESCRIPTION
**Pull request by rtetley**
Loosely related to #18694.
I need to get a loc for the errors raised by a require for a lib compiled with the wrong coq version.
This is an attempt at fixing this, I need reviews and advice on how best proceed.

### CODE REVIEW

File: lib/system.ml [hunk 1]
```diff
@@ -248,17 +248,17 @@ type magic_number_error = {filename: string; actual: int32; expected: int32}
 exception Bad_magic_number of magic_number_error
 exception Bad_version_number of magic_number_error
 
-let with_magic_number_check f a =
+let with_magic_number_check ?loc f a =
   try f a
   with
   | Bad_magic_number {filename=fname; actual; expected} ->
-    CErrors.user_err
+    CErrors.user_err ?loc:loc
```
Comment (SkySkimmer) [inline lib/system.ml:255]:
```suggestion
    CErrors.user_err ?loc
```
same for all other occurrences

File: vernac/library.ml [hunk 2]
```diff
@@ -267,10 +267,10 @@ let universes_seg : seg_univ option ObjFile.id = ObjFile.make_id "universes"
 let tasks_seg () : (Opaqueproof.opaque_handle option, 'doc) tasks option ObjFile.id = ObjFile.make_id "tasks"
 let opaques_seg : seg_proofs ObjFile.id = ObjFile.make_id "opaques"
 
-let intern_from_file lib_resolver dir =
+let intern_from_file lib_resolver dir loc =
```
Comment (SkySkimmer) [inline vernac/library.ml:270]:
```suggestion
let intern_from_file ?loc lib_resolver dir =
```

File: vernac/library.mli [hunk 3]
```diff
@@ -27,7 +27,7 @@ val require_library_from_dirpath : library_t list -> unit
 
 val require_library_syntax_from_dirpath
   :  lib_resolver:(DirPath.t -> CUnix.physical_path)
-  -> DirPath.t list
+  -> (DirPath.t * Libnames.qualid) list
```
Comment (SkySkimmer) [inline vernac/library.mli:30]: I would do
```suggestion
  -> DirPath.t Loc.located list
```
if there's no need for the qualid.

File: vernac/synterp.ml [hunk 4]
```diff
@@ -292,20 +292,20 @@ let synterp_require from export qidl =
...
   let filenames = Library.require_library_syntax_from_dirpath ~lib_resolver modrefl in
   Option.iter (fun (export,cats) ->
       let cats = synterp_import_cats cats in
-      List.iter2 (fun m (_,f) ->
+      List.iter2 (fun (_, m) (_, f) ->
           import_module_syntax_with_filter ~export cats (MPfile m) f)
         modrefl qidl)
     export;
-    filenames, modrefl
+    filenames, List.map (fun (_, d) -> d) modrefl
```
Comment (ejgallego) [inline vernac/synterp.ml:308]: `List.map snd` is a bit more brief.

File: vernac/synterp.ml [hunk 5]
```diff
@@ -292,20 +292,20 @@ let synterp_require from export qidl =
   let locate (qid,_) =
     let open Loadpath in
     match locate_qualified_library ?root qid with
-    | Ok (dir,_) -> dir
-    | Error LibUnmappedDir -> raise (UnmappedLibrary (root, qid))
-    | Error LibNotFound -> raise (NotFoundLibrary (root, qid))
+    | Ok (dir,_) -> (qid.loc, dir)
```
Comment (ejgallego) [inline vernac/synterp.ml:295]: As you are manipulating a `CAst.t`, you could instead try to preserve the `Ast` attributes for the client downwards, use for example `CAst.map_with_loc`.

Not a big deal tho.

I would be good at some point to remove the internal raising there, and use `Result.t`.

File: test-suite/output/bug_15097.out [hunk 6]
```diff
@@ -1,7 +1,6 @@
...
-File "./output/bug_15097.v", line 2, characters 0-44:
-The command has indeed failed with message:
-Cannot find a physical path bound to logical path
-Does.Not.Exist with prefix Coq.
+File not found: theories/Init/Prelude.vo
+The path for Coq libraries is wrong.
+Coq libraries are shipped in the coq-stdlib package.
+Please check the COQLIB env variable or the -coqlib option.
+
+coqc exited with code 1
```
Comment (ejgallego) [inline test-suite/output/bug_15097.out:6]: This seems strange?

Comment (rtetley) [reply test-suite/output/bug_15097.out]: Should be fixed now ?

**Comment by rtetley** [comment]
@coqbot run full CI

**Comment by rtetley** [comment]
@coqbot run full CI

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**